flink-clients/src/test/java/org/apache/flink/client/deployment/application/FromClasspathEntryClassInformationProviderTest.java

```diff
@@ -0,0 +1,173 @@
+    }
+
+    @Test(expected = NullPointerException.class)
+    public void testJobClassOnUserClasspathWithMissingUserClasspath() throws FlinkException {
+        FromClasspathEntryClassInformationProvider.create("jobClassName", null);
+    }
+
+    @Test
+    public void testJobClassOnUserClasspathWithoutExplicitJobClassName() throws IOException {
+        FromClasspathEntryClassInformationProvider testInstance =
+                FromClasspathEntryClassInformationProvider.createFromClasspath(
+                        classpathProvider.getURLUserClasspathWithEntryClass());
+
+        assertThat(testInstance.getJobClassName().isPresent(), is(true));
+        assertThat(testInstance.getJobClassName().get(), is(classpathProvider.getJobClassName()));
+        assertThat(testInstance.getJarFile().isPresent(), is(false));
+    }
+
+    @Test(expected = NoSuchElementException.class)
+    public void testMissingJobClassOnUserClasspathWithoutExplicitJobClassName() throws IOException {
+        FromClasspathEntryClassInformationProvider.createFromClasspath(
+                classpathProvider.getURLUserClasspathWithoutEntryClass());
+    }
+
+    @Test(expected = IllegalArgumentException.class)
+    public void testTooManyMainMethodsOnUserClasspath() throws IOException {
+        FromClasspathEntryClassInformationProvider.createFromClasspath(
+                classpathProvider.getURLUserClasspathWithTwoEntryClasses());
+    }
+
+    @Test(expected = NullPointerException.class)
+    public void testJobClassOnUserClasspathWithoutExplicitJobClassNameAndMissingUserClasspath()
+            throws IOException {
+        FromClasspathEntryClassInformationProvider.createFromClasspath(null);
+    }
+
+    @Test
+    public void testJobClassNameFromSystemClasspath() throws IOException {
+        classpathProvider.setSystemClasspathWithEntryClass();
+        FromClasspathEntryClassInformationProvider testInstance =
+                FromClasspathEntryClassInformationProvider.createFromSystemClasspath();
+        assertThat(testInstance.getJobClassName().isPresent(), is(true));
+        assertThat(testInstance.getJobClassName().get(), is(classpathProvider.getJobClassName()));
+        assertThat(testInstance.getJarFile().isPresent(), is(false));
+    }
+
+    @Test(expected = NoSuchElementException.class)
+    public void testMissingJobClassNameFromSystemClasspath() throws IOException {
+        classpathProvider.setSystemClasspathWithoutEntryClass();
+        FromClasspathEntryClassInformationProvider.createFromSystemClasspath();
+    }
+
+    @Test(expected = IllegalArgumentException.class)
+    public void testTooManyMainMethodsOnSystemClasspath() throws IOException {
+        classpathProvider.setSystemClasspathWithTwoEntryClasses();
+        FromClasspathEntryClassInformationProvider.createFromSystemClasspath();
+    }
+
+    @Test
+    public void testJarFromSystemClasspathSanityCheck() {
```

Comment:
what do we need this test for? If you want to check that we can properly load something from the system CP without any modifications to it, then why not put a jar on the test classpath and try loading a job from that.

Reply:
That's a sanity check I copied over from the old `ClassPathPackagedProgramRetrieverTest`. I thought it's reasonable to check whether the jar loading without modifying the `java.class.path` System property. Isn't that essentially the same as what you suggest? ...just with less code?

Reply:
We can keep it because it already exists, but what I objected to is searching for the junit jar on a conceptual level. It is not sound to couple the usage of some testing library to the correctness of production code.

Reply:
You have a point. I agree. But I would also vote for keeping it like this for now as I want to avoid opening another refactoring sub-project.